File: app/src/test/java/com/memoizlabs/kotlin_playground/currency/CurrencyMapperTest.java
package com.memoizlabs.kotlin_playground.currency;

import com.squareup.okhttp.mockwebserver.MockWebServer;

import org.junit.Test;

import java.math.BigDecimal;

import rx.observers.TestSubscriber;

public class CurrencyMapperTest {

    private String responseJson = "{\"base\":\"GBP\",\"date\":\"2016-07-01\",\"rates\":{\"USD\":1.3283,\"EUR\":1.4322}}";
    private GBPExchangeRate expectedGBPExchangeRate = new GBPExchangeRate(new BigDecimal(1.3283), new BigDecimal(1.4322));

    @Test
    public void fetchesExchangeRates() {
        MockWebServer mockWebServer = MockWebServerBuilder.aMockWebServer().returningJson(responseJson).start();
        CurrencyMapper currencyMapper = new CurrencyMapper(RetrofitUtil.createService(ExchangeRetrofitService.class, mockWebServer.url("/").toString()));

        TestSubscriber<GBPExchangeRate> testSubscriber = TestSubscriber.create();
        currencyMapper.getRate().subscribe(testSubscriber);

        testSubscriber.assertNoErrors();
        testSubscriber.assertValue(expectedGBPExchangeRate);
    }

}

File: app/src/test/java/com/memoizlabs/kotlin_playground/currency/CurrencyMapperTest.kt
package com.memoizlabs.kotlin_playground.currency

import org.junit.Test
import rx.observers.TestSubscriber

class CurrencyMapperTest {

    private val responseJson = "{\"base\":\"GBP\",\"date\":\"2016-07-01\",\"rates\":{\"USD\":1.3283,\"EUR\":1.4322}}"
    private val expectedGBPExchangeRate = GBPExchangeRate(Money(1.3283), Money(1.4322))

    @Test
    fun fetchesExchangeRates() {
        val mockWebServer = MockWebServerBuilder.aMockWebServer().returningJson(responseJson).start()
        val currencyMapper = CurrencyMapper(createService(ExchangeRetrofitService::class.java, mockWebServer.url("/").toString()))

        val testSubscriber = TestSubscriber.create<GBPExchangeRate>()
        currencyMapper.getRate().subscribe(testSubscriber)

        testSubscriber.assertNoErrors()
        testSubscriber.assertValue(expectedGBPExchangeRate)
    }

}